server/controllers/articles.js

```diff
@@ -317,5 +320,206 @@ class Articles {
+    };
+
+    const result = await Article.findAll({
+      order: [['createdAt', 'DESC']],
+      limit: pagination.limit,
+      offset: pagination.offset,
+      include: [
+        {
+          model: User,
+          attributes: ['username', 'email', 'image'],
+          where,
+        },
+        {
+          model: Tags,
+        },
+        {
+          model: Rating,
+          attributes: ['rate']
+        }
+      ]
+    });
+
+    return res.status(200).json({
+      status: 200,
+      data: { articles: result }
+    });
+  }
+
+  /**
+   * helps filter by title
+   *
+   * @description Check whether title matches or
+   * if title contains the value passed as params
+   * @static
+   * @param {object} res
+   * @param {string} title
+   * @param {object} pagination
+   * @returns {object} response
+   * @memberof Articles
+  */
+  static async filterByTitle(res, title, pagination) {
+    const where = {
+      title: { [Op.iLike]: `%${title}%` }
+    };
+
+    const result = await Article.findAll({
+      order: [['createdAt', 'DESC']],
+      limit: pagination.limit,
+      offset: pagination.offset,
+      where,
+      include: includeQuery
+    });
+
+    return res.status(200).json({
+      status: 200,
+      data: { articles: result }
+    });
+  }
+
+  /**
+   * helps filter by keyword
+   *
+   * @description search the keyword inside the article title, body or description
+   * @static
+   * @param {object} res
+   * @param {string} keywords
+   * @param {object} pagination
+   * @returns {object} response
+   * @memberof Articles
+  */
+  static async filterByKeywords(res, keywords, pagination) {
+    const where = {
+      [Op.or]: [
+        { title: { [Op.iLike]: `%${keywords}%` } },
+        { body: { [Op.iLike]: `%${keywords}%` } },
+        { description: { [Op.iLike]: `%${keywords}%` } }
+      ]
+    };
+
+    const result = await Article.findAll({
+      order: [['createdAt', 'DESC']],
+      limit: pagination.limit,
+      offset: pagination.offset,
+      where,
+      include: includeQuery
+    });
+
+    return res.status(200).json({
+      status: 200,
+      data: { articles: result }
+    });
+  }
+
+  /**
+   * helps filter by tag
+   *
+   * @description search if tagName contains tags pass in params
+   * @static
+   * @param {object} res
+   * @param {string[]} tags
+   * @returns {object} response
+   * @memberof Articles
+  */
+  static async filterByTags(res, tags) {
+    const tagsList = Articles.tagsToArray(tags);
+    const where = {
+      [Op.or]: [
+        { tagName: tagsList }
+      ]
+    };
+
+    const result = await Tags.findAll({
+      where,
+      include: [{
+        model: Article,
+        include: [
+          {
+            model: User,
+            attributes: ['username', 'email', 'image'],
+          },
+          {
+            model: Rating,
+            attributes: ['rate']
+          }
+        ],
+        order: [['createdAt', 'DESC']],
+      }]
+    });
+
```

Comment:
What if we respond with an appropriate message when the result doesnt return anything instead of returning an empty array?

Reply:
> What if we respond with an appropriate message when the result doesn't return anything instead of returning an empty array? can also apply to article title and author

@dusmel, I think the frontend should handle this case as we agreed on using `infinite scrolling`. The backend should not respond with a message. An empty array means `no data to display`. In brief, the frontend will display a message in case the article array is empty